src/rpc/client.cpp

```diff
@@ -214,15 +214,25 @@ CRPCConvertTable::CRPCConvertTable()
 
 static CRPCConvertTable rpcCvtTable;
 
+static bool IsHexDigit(const char c) {
+    return ('0' <= c && c <= '9') || ('a' <= c && c <= 'f') || ('A' <= c && c <= 'F');
+}
+
+static bool IsHashVal(const std::string& str) {
+    return (str.length() == 64) && std::all_of(str.begin(), str.end(), IsHexDigit);
+}
+
 /** Non-RFC4627 JSON parser, accepts internal values (such as numbers, true, false, null)
  * as well as objects and arrays.
  */
 UniValue ParseNonRFCJSONValue(const std::string& strVal)
 {
     UniValue jVal;
     if (!jVal.read(std::string("[")+strVal+std::string("]")) ||
```

Comment:
```suggestion
    if ((!jVal.read(std::string("[")+strVal+std::string("]")) &&
        !jVal.read(std::string("[\"")+strVal+std::string("\"]"))) ||
```
Just as an idea, feel free to ignore :)
This would be shorter, requires no additional methods and also does the job (except for `getblockstats "'<blockhash>'"`). But - and that could be more important - it keeps the error message if the blockhash is e.g. too long (`hash_or_height must be of length 64 (not 66, for ...`, while your solution throws an JSON parsing error which could be misleading for a user 🤔

Reply:
If I see that correctly what this change will do is it will end up passing everything as a string to the server and never throw that error client-side. So basically the same as #15448 which removes `ParseNonRFCJSONValue` and only does validation server-side (the arguably better error message is coming from the server). The feedback to that was that the validation shouldn't be removed completely from the client-side which is the approach I am trying here.

Reply:
Sounds reasonable 👍